### PR TITLE
Complete bcp handling

### PR DESCRIPTION
**Pull request by bratseth**
- Add bcp to schema validation
- Read endpoints defined in bcp elements

### CODE REVIEW

File: config-model-api/src/main/java/com/yahoo/config/application/api/xml/DeploymentSpecXmlReader.java [hunk 1]
```diff
@@ -329,150 +344,184 @@ private void readEndpoints(Element parent, Optional<String> instance, List<Step>
...
-                        illegal(msgPrefix + "duplicate 'region' element: '" + region + "'");
-                }
+    /**
+     * @param parentElement
+     * @param endpointElement
+     * @param level decide what this method is reading TODO: Split into different methods instead
+     * @param instance the instance this applies to, or empty if it does not apply to an instance (application endpoints)
+     * @param steps
+     * @param forRegions the regions this applies to (for bcp), or empty (otherwise) to read this from "region" subelements
+     * @param endpointsByZone a map containing any zone endpoints rtead by this
```
Comment (mpolden) [inline config-model-api/src/main/java/com/yahoo/config/application/api/xml/DeploymentSpecXmlReader.java:363]: rtead -> read

File: config-model-api/src/main/java/com/yahoo/config/application/api/xml/DeploymentSpecXmlReader.java [hunk 2]
```diff
@@ -329,150 +344,184 @@ private void readEndpoints(Element parent, Optional<String> instance, List<Step>
...
+     * @return the endpoint read, unless it is added to endspointsByZone instead *sob*
+     */
+    static Optional<Endpoint> readEndpoint(Element parentElement,
+                                           Element endpointElement,
+                                           Endpoint.Level level,
+                                           Optional<String> instance,
+                                           List<Step> steps,
+                                           Collection<RegionName> forRegions,
+                                           Map<String, Map<RegionName, List<ZoneEndpoint>>> endpointsByZone) {
```
Comment (mpolden) [inline config-model-api/src/main/java/com/yahoo/config/application/api/xml/DeploymentSpecXmlReader.java:372]: :see_no_evil:

File: config-model-api/src/test/java/com/yahoo/config/application/api/DeploymentSpecWithBcpTest.java [hunk 3]
```diff
@@ -241,6 +244,164 @@ public void fractionalMembershipValidation() {
...
+        assertEquals("east", mainEndpoints.get(0).endpointId());
+        assertEquals(List.of(RegionName.from("us-east1"), RegionName.from("us-east2"), RegionName.from("us-central1")),
+                     mainEndpoints.get(0).regions());
+        assertEquals("west", mainEndpoints.get(1).endpointId());
+        assertEquals(List.of(RegionName.from("us-west1"), RegionName.from("us-west2"), RegionName.from("us-central1")),
+                     mainEndpoints.get(1).regions());
+    }
+
+    @Test
+    public void endpontsDefinedInBcpValidation1() {
```
Comment (mpolden) [inline config-model-api/src/test/java/com/yahoo/config/application/api/DeploymentSpecWithBcpTest.java:350]: endponts -> endpoints

File: config-model-api/src/test/java/com/yahoo/config/application/api/DeploymentSpecWithBcpTest.java [hunk 4]
```diff
@@ -241,6 +244,164 @@ public void fractionalMembershipValidation() {
...
+        try {
+            DeploymentSpec.fromXml(r);
+        }
+        catch (IllegalArgumentException e) {
+            assertEquals("The default <bcp> element at the root cannot define endpoints", Exceptions.toMessageString(e));
+        }
+    }
+
+    @Test
+    public void endpontsDefinedInBcpValidation2() {
```
Comment (mpolden) [inline config-model-api/src/test/java/com/yahoo/config/application/api/DeploymentSpecWithBcpTest.java:377]: endponts -> endpoints

**Review by mpolden — APPROVED** [review]
:+1: 

I'm out next week.

File: config-model-api/src/main/java/com/yahoo/config/application/api/xml/DeploymentSpecXmlReader.java (hunk 2 above)
Comment (bratseth) [reply config-model-api/src/main/java/com/yahoo/config/application/api/xml/DeploymentSpecXmlReader.java:372]: Yes :-)

**Pull request merged**